cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -135,6 +135,69 @@ using rhs_iterator = strong_index_iterator<rhs_index_type>;
 
 namespace lexicographic {
 
+/**
+ * @brief Relational comparator functor that evaluates `NaN` as not less, not equivalent, and not
+ * greater than all other values. This is IEE 754 compliant.
+ */
+struct physical_element_comparator {
+  /**
+   * @brief Shared operator for all `Element` types relational comparisons.
+   *
+   * @param[in] lhs first element
+   * @param[in] rhs second element
+   * @return Indicates the relationship between the elements in
+   * the `lhs` and `rhs` columns.
+   */
+  template <typename Element>
+  __device__ weak_ordering operator()(Element lhs, Element rhs)
+  {
+    return detail::compare_elements(lhs, rhs);
+  }
+};
+
+/**
+ * @brief Relational comparator functor that evaluates `NaN` as equivalent to other `NaN`s and
+ * greater than all other values.
+ */
+struct sorting_physical_element_comparator : physical_element_comparator {
+  /**
+   * @brief A specialization for non-floating-point `Element` type relational
+   * comparison to derive the order of the elements with respect to `lhs`.
+   *
+   * @param[in] lhs first element
+   * @param[in] rhs second element
+   * @return Indicates the relationship between the elements in
+   * the `lhs` and `rhs` columns.
+   */
+  template <typename Element, std::enable_if_t<not std::is_floating_point_v<Element>>* = nullptr>
+  __device__ weak_ordering operator()(Element lhs, Element rhs)
+  {
+    return detail::compare_elements(lhs, rhs);
+  }
+
+  /**
+   * @brief A specialization for floating-point `Element` type relational comparison that compares
+   * the nulls according to null order.
+   *
+   * @param lhs_is_null boolean representing if lhs is null
+   * @param rhs_is_null boolean representing if lhs is null
+   * @return Indicates the relationship between null in lhs and rhs columns.
+   */
+  template <typename Element, std::enable_if_t<std::is_floating_point_v<Element>>* = nullptr>
+  __device__ weak_ordering operator()(Element lhs, Element rhs)
```

Comment:
Do this throughout. It occasionally helps with making sure the compiler boils everything away. 
```suggestion
  constexpr __device__ weak_ordering operator()(Element const lhs, Element const rhs) const noexcept
```

Reply:
Do we do `constexpr` left or right of device? I see both, but mostly `__device__ constexpr`